xml-conduit/Text/XML/Stream/Render.hs

```diff
@@ -115,32 +117,53 @@ orderAttrs orderSpec = order
 -- the blaze-builder package, and allow the create of optimally sized
 -- 'ByteString's with minimal buffer copying.
 renderBuilder :: Monad m => RenderSettings -> Conduit Event m Builder
-renderBuilder RenderSettings { rsPretty = True, rsNamespaces = n, rsUseCDATA = useCDATA } = prettify =$= renderBuilder' n True useCDATA
-renderBuilder RenderSettings { rsPretty = False, rsNamespaces = n, rsUseCDATA = useCDATA } = renderBuilder' n False useCDATA
+renderBuilder settings@(RenderSettings { rsPretty = True }) = prettify =$= renderBuilder' settings
+renderBuilder settings@(RenderSettings { rsPretty = False }) = renderBuilder' settings
 
-renderBuilder' :: Monad m => [(Text, Text)] -> Bool -> (Content -> Bool) -> Conduit Event m Builder
-renderBuilder' namespaces0 isPretty useCDATA = do
+-- | Same as 'renderBuilder' but allows for flushing output when needed.
+--   Because it's guaranteed that 'Flush' would emit all the previously received data,
+--   this function doesn't support prettifying. Also it won't do certain other
+--   optimizations, like using self-closed tags.
+renderBuilderFlush :: Monad m => RenderSettings -> Conduit (Flush Event) m (Flush Builder)
+renderBuilderFlush settings@(RenderSettings { rsPretty = True }) = error "renderBuilderFlush: prettifying is not supported"
```

Comment:
Sorry, I'm bad at comments and explanations -_-

It was made this way to guarantee that Flush would emit all piped data further, acting as a barrier. Consider this scenario: I first send `Chunk EventBeginDocument{}`, then `Chunk EventBeginElement{} ` and then `Flush`. At this point I expect that output would consist of prologue and an opening tag. But if something relies on peeking, it can't be done because we would need the next Chunk to get event out of it. Hence, disabled everything that relies on peeking. This actually happens in XMPP where I need to send an opening tag and then wait for the server to response before sending next messages. 

On the other hand, while writing this I realized that it _can_ actually be done if I would peek at the Flush stream directly and decide conservatively if Flush is encountered. For example, in the case of `isClosed` I can just default it to False when that happens. This would get me a nice optimization, too (less text to send for self-closing tags). Stay tuned for an updated version of the patch! 

On May 10, 2016 1:08:24 PM GMT+03:00, Michael Snoyman <notifications@github.com> wrote:
>>  
>> -renderBuilder' :: Monad m => [(Text, Text)] -> Bool -> (Content ->
>Bool) -> Conduit Event m Builder
>> -renderBuilder' namespaces0 isPretty useCDATA = do
>> +-- | Same as 'renderBuilder' but allows for flushing output when
>needed.
>> +--   Because it's guaranteed that 'Flush' would emit all the
>previously received data,
>> +--   this function doesn't support prettifying. Also it won't do
>certain other
>> +--   optimizations, like using self-closed tags.
>> +renderBuilderFlush :: Monad m => RenderSettings -> Conduit (Flush
>Event) m (Flush Builder)
>> +renderBuilderFlush settings@(RenderSettings { rsPretty = True }) =
>error "renderBuilderFlush: prettifying is not supported"
>
>Never mind, I read the comment. I see that there's a reason, I guess
>I'm just not following it. In any event, I think adding a comment
>explaining that `rsPretty` is ignored in this function would be
>preferable to errorring out.
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/snoyberg/xml/pull/85/files/efe4c6778b8e52bc03444d5ec6e2a29602910bb8#r62644585

-- 
Nikolay.